### DISCUSSION TITLE
How to not bundle react and react-dom the proper way ?

### DISCUSSION THREAD

**Discussion by gerardnico** [parcel-bundler/parcel#7322] [Q&A] [answered]
Hallo,

## The problem 
I try to not bundle react and react-dom but in my functional config, I get always this message.
```
@parcel/transformer-babel: @babel/preset-env does not support Parcel's targets, which will likely result in
unnecessary transpilation and larger bundle sizes.

  D:\temp\parcelreact\.babelrc:4:7
    3 |     [
  > 4 |       "@babel/preset-env",
  >   |       ^^^^^^^^^^^^^^^^^^^
    5 |       {
    6 |         "targets": {"node": "current"}

  ℹ Either remove @babel/preset-env to use Parcel's builtin transpilation, or replace with
     @parcel/babel-preset-env
  ℹ Learn more: https://parceljs.org/languages/javascript/#custom-plugins
```

## The context
I'm working on a php application where we render on several place React component.
The page has therefore already react and react-dom included and we don't want to bundle them again.

## Step 1 - Alias creation

As stated in the [doc](https://parceljs.org/features/dependency-resolution/#global-aliases), I have added to `package.json`, the aliases
```json
"alias": {
    "react-dom": {
      "global": "ReactDOM"
    },
    "react": {
      "global": "React"
    }
}
```

## Step 1 - Error: Jsx transpilation does not work with alias
Unfortunately, if i do that the babel transpilation does not work anymore and I get this error:
```
Uncaught TypeError: _jsxRuntime.jsx is not a function
```

## Step 2 - Adding Babel

I create a `.babelrc` file to do the jsx transpilation

```json
{
  "presets": [
    "@babel/preset-env",
    "@babel/preset-react"
  ]
}
```

## Step 2 - Final Error
It works but I got the below error/warning. The refresh is also working for every build.

```
@parcel/transformer-babel: @babel/preset-env does not support Parcel's targets, which will likely result in
unnecessary transpilation and larger bundle sizes.

  D:\temp\parcelreact\.babelrc:4:7
    3 |     [
  > 4 |       "@babel/preset-env",
  >   |       ^^^^^^^^^^^^^^^^^^^
    5 |       {
    6 |         "targets": {"node": "current"}

  ℹ Either remove @babel/preset-env to use Parcel's builtin transpilation, or replace with
     @parcel/babel-preset-env
  ℹ Learn more: https://parceljs.org/languages/javascript/#custom-plugins
```

## Other solutions: includeNodeModules
I tried to not [include the react modules in the target](https://parceljs.org/features/targets/#includenodemodules)

```json
"includeNodeModules": {
   "react": false
}
```

But there is no transpilation anymore and I got the same behavior than with the aliases
```
Uncaught TypeError: _jsxRuntime.jsx is not a function
```

## Conclusion

I'm working now with alias and babel but parcel is not happy.

Did I miss something ?

Thanks

## Extra

FYI, the html file that I use is this one and is just derived from the [React documentation](https://parceljs.org/recipes/react/) where React and React-dom are included.

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8" />
    <title>My Parcel App</title>
    <script src="https://unpkg.com/react@17/umd/react.development.js" crossorigin></script>
    <script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js" crossorigin></script>
</head>
<body>
<div id="app"></div>
<script type="module" src="index.js"></script>
</body>
</html>
```

And package.json
```json
{
  "name": "parcelreact",
  "version": "1.0.0",
  "source": "src/react/index.html",
  "alias": {
    "react-dom": {
      "global": "ReactDOM"
    },
    "react": {
      "global": "React"
    }
  },
  "devDependencies": {
    "@babel/core": "^7.16.0",
    "@babel/preset-env": "^7.16.4",
    "@babel/preset-react": "^7.16.0",
    "@types/react": "^17.0.35",
    "@types/react-dom": "^17.0.11",
    "parcel": "^2.0.1",
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  },
  "peerDependencies": {
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  }
}
```

**Comment by nickcernis** [comment, marked as answer]
I had the same issue where using the aliases you mention for `react-dom` and `react` would result in `Uncaught TypeError: _jsxRuntime.jsx is not a function` in the browser console when using Parcel 2.0.1 to transpile JS/JSX.

I was able to resolve it by:

1. Creating a `.parcelrc` with the content below to ensure Parcel is using its own transpilation instead of Babel's. (I still needed Babel separately in the project for Jest — if you don't need Babel for anything else you can likely skip this step and remove your Babel deps to use Parcel's transpilation.). This is documented at https://parceljs.org/languages/javascript/#usage-with-other-tools.

```
{
	"extends": "@parcel/config-default",
	"transformers": {
		"*.{js,mjs,jsx,cjs,ts,tsx}": [
			"@parcel/transformer-js",
			"@parcel/transformer-react-refresh-wrap"
		]
	}
}
```

2. Removing `react` and `react-dom` from dependencies in `package.json` (they should not be needed because you're aliasing them and including them externally via the links in the HTML). It seems that Parcel will use react and react-dom from node_modules first if they're present and ignore the alias — it looks like you have to remove the dependencies for the alias to take effect.

**Reply by gerardnico** [reply]
Thanks. I had already switched to webpack. I can't test unfortunately.

**Reply by joelash** [reply]
@nickcernis I am testing this and am not having luck with this solution. Is it still working for you? @gerardnico what was the lift like to switch to webpack? I'm debating if I need to go that way.

**Reply by nickcernis** [reply]
We ended up reinstating the `react` and `react-dom` dependencies in `package.json`, and it still works for us with the above `.parcelrc` and aliases (we're currently on parcel 2.3.1).

**Reply by mischnic** [reply]
The root problem appears to be that the [automatic react runtime](https://play.swc.rs/?version=1.2.245&code=H4sIAAAAAAAAA0vOzyvOz0nVy8lP17BJySzTt9O0BgD83VisFAAAAA%3D%3D&config=H4sIAAAAAAAAAy2MSw6CUAxF99I4JKAOTGQPLqKpFR7xfdL2RQlh7xZwdnvv6VngpDRyROhhNCvad92kObVHq5aF2yxDpx8SggYmJegXKCjKsiWdk%2BHX35mcJwnFdswrk8prA4YysG2EXs%2BXu68mmPSVJW4CYSTbQ00WIjuI1XJECwTr6oKgj%2Fysbz6MDcT%2FtYDNhXfxzckfbCK5rswAAAA%3D) imports `react/jsx-runtime` which apparently doesn't work well together with aliasing `react` to a global variable.

**Reply by gerardnico** [reply]
@joelash  With webpack, you do it with the [externals](https://datacadamia.com/web/build/webpack/dependency#exclude) property. Easy-peasy.
